src/typechecker.erl

```diff
@@ -3563,39 +3563,30 @@ add_type_pat(Expr={op, P, _Op, _Pat1, _Pat2}, Ty, TEnv, VEnv) ->
     %% the pattern has the correct type) but we also need to check that the left
     %% argument is not smaller than the right argument. But instead of implementing
     %% such a check, we simply evaluate the pattern as an expression.
-    {value, Val, _} = erl_eval:expr(Expr, orddict:new()),
-    {PatTy, ValType} = if is_integer(Val) ->
-                              {{integer, erl_anno:new(0), Val},
-                               {integer, erl_anno:new(0), Val}};
-                          is_float(Val) ->
-                              {type(none), type(float)}
-                       end,
-    case subtype(ValType, Ty, TEnv) of
-        {true, Cs} ->
-            {PatTy, ValType, VEnv, Cs};
-        false ->
-            throw({type_error, operator_pattern, P, Expr, Ty})
-    end;
-add_type_pat(Expr={op, P, _Op, _Pat}, Ty, TEnv, VEnv) ->
-    {value, Val, _} = erl_eval:expr(Expr, orddict:new()),
-    {PatTy, ValType} = if is_integer(Val) ->
-                              {{integer, erl_anno:new(0), Val},
-                               {integer, erl_anno:new(0), Val}};
-                          is_float(Val) ->
-                              {type(none), type(float)}
-                       end,
-    case subtype(ValType, Ty, TEnv) of
-        {true, Cs} ->
-            {PatTy, ValType, VEnv, Cs};
-        false ->
-            throw({type_error, operator_pattern, P, Expr, Ty})
-    end;
+    add_type_pat_literal(OpPat, Ty, TEnv, VEnv);
+add_type_pat(OpPat = {op, _Anno, _Op, _Pat}, Ty, TEnv, VEnv) ->
+    add_type_pat_literal(OpPat, Ty, TEnv, VEnv);
 add_type_pat(Pat, {ann_type, _, [_, Ty]}, TEnv, VEnv) ->
     add_type_pat(Pat, Ty, TEnv, VEnv);
 
 add_type_pat(Pat, Ty, _TEnv, _VEnv) ->
     throw({type_error, pattern, element(2, Pat), Pat, Ty}).
 
+add_type_pat_literal(Pat, Ty, TEnv, VEnv) ->
+    case erl_eval:partial_eval(Pat) of
+        Pat ->
+            %% illegal, non-constant pattern - this should not happen
+            throw({illegal_pattern, Pat});
```

Comment:
Should not happen for what reason? I guess only hand-coded ASTs can include this. Perhaps we don't need this clause but instead just `Literal when Literal =/= Pat` in the other clause? Then `Pat` gives a case_clause error and code coverage can increase. (I assume this is an uncovered clause.)

Reply:
I was thinking about exactly the same when I saw codecov complaining. However at the undefined-records (somewhere in this thread https://github.com/josefs/Gradualizer/pull/147#issuecomment-457855199) we agreed that we assume the least from the AST and (although proactively don't look for such errors) when we hit it, we show a meaningful message.

So I will just add a little test to cover this case similar to this https://github.com/josefs/Gradualizer/blob/master/test/typechecker_tests.erl#L546

Reply:
Sure, if you add a test for this, at least it's not completely dead code. I do think there's a difference to undefined records though.

For the undefined records, it's still valid syntax, so it's still a realistic scenario to hit when we're type checking source files just using the parser, which is something we do support. An AST that the parser cannot produce is not the same; it's not realistic IMO.

We were talking about calling the compiler for the validation step when checking source files, which would catch undefined records, but we didn't implement it yet, did we? If we do, I think we don't need the code for handling undefined records either.

Reply:
Actually the parser successfully parses this illegal patter (see the added testcase) only the linter catches it.
If gradualizer will ever support running as a parse_transform (to plug it straight into the compilation pipeline) that is the only scenario when we might receive such unchecked AST, and I think in that case we cannot call out to the compiler/linter. This use case (which we might decide to never implement for some reason or the other) stopped me to implement the erl_lint check for source code, and convinced me that we need to handle every case the parser can possibly spit out.